tests/integration_async/test_workflows.py

```
@@ -0,0 +1,384 @@
+# (at your option) any later version.
+#
+# INSPIRE is distributed in the hope that it will be useful,
+# but WITHOUT ANY WARRANTY; without even the implied warranty of
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+# GNU General Public License for more details.
+#
+# You should have received a copy of the GNU General Public License
+# along with INSPIRE. If not, see <http://www.gnu.org/licenses/>.
+#
+# In applying this license, CERN does not waive the privileges and immunities
+# granted to it by virtue of its status as an Intergovernmental Organization
+# or submit itself to any jurisdiction.
+
+from __future__ import absolute_import, division, print_function
+
+import json
+import os
+
+import pkg_resources
+import time
+
+from datetime import datetime, timedelta, date
+
+from copy import deepcopy
+
+from invenio_db import db
+from invenio_search import current_search_client as es
+from invenio_workflows import ObjectStatus, start, workflow_object_class
+
+
+def generate_record():
+    """Provide record fixture."""
+    json_data = json.loads(pkg_resources.resource_string(
+        __name__,
+        os.path.join(
+            '../integration/workflows/fixtures',
+            'oai_arxiv_core_record.json'
+        )
+    ))
+
+    if 'preprint_date' in json_data:
+        json_data['preprint_date'] = date.today().isoformat()
+
+    return json_data
+
+
+def check_wf_state(workflow_id, desired_status, max_time=600):
+    """Waits for the workflow to go to desired status
+    Args:
+        workflow: workflow to check
+        desired_state: desired state
+        max_time: maximum time to wait in seconds, otherwise raise exception
+
+    Returns: None
+
+    """
+    start = datetime.now()
+    end = start + timedelta(seconds=max_time)
+    while (True):
+        db.session.close()
+        if workflow_object_class.get(workflow_id).status == desired_status:
+            return
+        if datetime.now() > end:
+            raise AssertionError
```

AssertionError isn't it a bit arbitrary? Maybe a custom exception would be easier to understand

---

In usual (not async) case I would just use `assert` here, so using `AssertionError` looks more natural and intuitive here.

---

``assert`` in general hides the real error, I'd still go for a custom error to understand why failed

---

Custom exception is too much here, but I've added a message which will explain more.